### PR TITLE
Replace Haproxy with Traefik

### PR DESCRIPTION
**Pull request by BrianHicks**
This PR adds Traefik (http://traefik.github.io) as the default proxy layer in Mantl. We've seen success with it using it with test clusters and internal and external traffic on some production clusters. Specifically, this PR:

- removes `haproxy` role from the code and docs
- adds `traefik` role to code and docs
- documents migrating from `haproxy` to `traefik`
- adds new "edge" role (thanks again for the idea, @eirslett!) allowing us to sidestep some Marathon authentication issues for now and keep a good segregation of security between workers and control nodes. Adds support for all clouds we support using this new role.
- fixes an invalid config in the AWS module (Amazon dropped support for our image using `m1.small`, so I've changed it to `m3.medium`.)
- adds a simple cloudflare module that I was using to test

I've tested this on GCE and lightly on AWS and DO, but the Terraform files need to be taken for a spin in Openstack (including floating IPs), softlayer, and vsphere.

Closes #773

### CODE REVIEW

**Comment by BrianHicks** [comment]
9356de0, 93096e1, and f4b5c9a reverted the commit and added deprecation warnings to that and the docs.

@stevendborrelli look good with that thought?

File: terraform/aws/aws.tf [hunk 1]
```diff
@@ -290,6 +324,26 @@ resource "aws_security_group" "ui" {
   }
 }
 
+resource "aws_security_group" "edge" {
```
Comment (ryane) [inline terraform/aws/aws.tf]: no ssh access to the edge nodes prevents ansible provisioning right now

File: terraform/gce/gce.tf [hunk 2]
```diff
@@ -31,18 +34,12 @@ resource "google_compute_firewall" "mi-firewall-external" {
     protocol = "icmp"
   }
 
-  # RDP
   allow {
     protocol = "tcp"
     ports = [
-      "22",   # SSH
```
Comment (ryane) [inline terraform/gce/gce.tf]: did we decide to close off mesos, marathon, and consul ports? if so, should we make AWS consistent?

Comment (BrianHicks) [inline terraform/gce/gce.tf]: ah no, we didn't. I guess this got committed by mistake.

File: terraform/aws/aws.tf (hunk 1 above)
Comment (BrianHicks) [inline terraform/aws/aws.tf]: whoops! I had made a partial change to make a "common" SG that included 22. I guess that's a bug.

**Comment by ryane** [comment]
during ansible run:

```
TASK: [docker | enable docker] ************************************************
changed: [resching-aws-control-02]
changed: [resching-aws-worker-003]
changed: [resching-aws-worker-001]
changed: [resching-aws-control-01]
changed: [resching-aws-worker-002]
changed: [resching-aws-worker-004]
changed: [resching-aws-control-03]
failed: [resching-aws-edge-02] => {"failed": true}
msg: Job for docker.service failed. See 'systemctl status docker.service' and 'journalctl -xn' for details.

failed: [resching-aws-edge-01] => {"failed": true}
msg: Job for docker.service failed. See 'systemctl status docker.service' and 'journalctl -xn' for details.
```

on edge node:

```
$ sudo systemctl status docker
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled)
   Active: failed (Result: timeout) since Tue 2015-11-17 19:06:59 UTC; 3min 1s ago
     Docs: http://docs.docker.com
 Main PID: 5132

Nov 17 19:05:48 ip-10-0-95-72 systemd[1]: Starting Docker Application Container Engine...
Nov 17 19:05:48 ip-10-0-95-72 docker[5132]: time="2015-11-17T19:05:48.891194622Z" level=error msg="WARNING: No --storage-opt dm.thinpooldev specified, using loopback; this configuration is strongly discouraged for production use"
Nov 17 19:05:48 ip-10-0-95-72 docker[5132]: time="2015-11-17T19:05:48.905799692Z" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)"
Nov 17 19:06:48 ip-10-0-95-72 systemd[1]: docker.service operation timed out. Terminating.
Nov 17 19:06:59 ip-10-0-95-72 systemd[1]: Failed to start Docker Application Container Engine.
Nov 17 19:06:59 ip-10-0-95-72 systemd[1]: Unit docker.service entered failed state.
```

any idea? do we need to add some additional conditionals since the edge node doesn't have an lvm volume?

**Comment by BrianHicks** [comment]
I had missed some things earlier, I thought the LVM stuff was only for gluster. Reading again, I'll need to add disks to those new resources. I'll get started on that. Thanks for your patience.

**Comment by BrianHicks** [comment]
@ryan give that a try, I fixed the errors (I think)

**Comment by ryane** [comment]
had a successful build on gce. no errors during terraform or ansible run and all services reporting as healthy. However, traefik did not seem to be proxying any services (404 for any request). Status:

```
$ sudo systemctl status traefik
traefik.service - Træfɪk, a modern reverse proxy
   Loaded: loaded (/etc/systemd/system/traefik.service; enabled)
   Active: active (running) since Tue 2015-11-17 21:57:30 UTC; 14min ago
     Docs: https://github.com/EmileVauge/traefik
 Main PID: 10830 (traefik)
   CGroup: /system.slice/traefik.service
           └─10830 /usr/bin/traefik /etc/traefik/traefik.toml

Nov 17 21:57:30 resching-gce-edge-01.c.asteris-mi.internal systemd[1]: Starting Træfɪk, a modern reverse proxy...
Nov 17 21:57:30 resching-gce-edge-01.c.asteris-mi.internal systemd[1]: Started Træfɪk, a modern reverse proxy.
Nov 17 21:57:35 resching-gce-edge-01.c.asteris-mi.internal traefik[10830]: time="2015-11-17T21:57:35Z" level=error msg="Failed to register for subscriptions, No cluster memebers available at the moment"
Nov 17 21:57:35 resching-gce-edge-01.c.asteris-mi.internal traefik[10830]: time="2015-11-17T21:57:35Z" level=error msg="Failed to create a client for marathon, error: No cluster memebers availab...the moment"
Hint: Some lines were ellipsized, use -l to show in full.
```

As soon as I restarted traefik on the edge nodes, the proxy was working as expected.

**Comment by BrianHicks** [comment]
The check in 5edab28 tests for that condition. I'm not sure where it's coming from.

**Comment by stevendborrelli** [comment]
Looks like it came up before marathon was running. I wonder if it should fail on conditions like that or keep retrying.

**Comment by BrianHicks** [comment]
It is scheduled to come up after Marathon for this reason (as of a9c0ba.) I think it actually is not being restarted correctly, I'm doing a little testing on a different way to configure it which I'll push ASAP.

**Comment by BrianHicks** [comment]
1882a5f solves the problem, nodes come up fine after Marathon now.

**Comment by ryane** [comment]
With the latest changes, just had another successful build on GCE but traefik-admin checks were failing on both edge nodes with: 

```
API contains no information. If this is not expected, try restarting traefik.
```

and proxies were 404'ing. Restarted traefik on both nodes. Health checks started passing and proxies were working correctly.

**Comment by ryane** [comment]
Saw the same exact scenario on a new AWS build

**Comment by BrianHicks** [comment]
Quite strange, what commit were you on? I just started up a GCE build and was fine. If you've customized your `terraform.yml`, make sure that the Traefik role is *after* the control role.

**Comment by ryane** [comment]
That was it. I was on the latest commit but I neglected to change the order of the roles. +1 on GCE

**Comment by ryane** [comment]
successful on AWS

**Comment by ryane** [comment]
+1 on openstack

**Comment by BrianHicks** [comment]
I tested on DO as well, and it worked fine there. That's our big 4 taken care of.

**Pull request merged**